A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,169 @@
+   endpoint IP address.
+
+## Background
+
+### Terminology
+
+* The *request hash key*, after being hashed, defines where a given request is
+  to be placed on the ring in order to find the closest endpoints.
+* The *endpoint hash key*, after being hashed, determines the locations of an
+  endpoint on the ring.
+
+### Using ring hash without xDS by explicitly setting the request hash key
+
+gRPC supports the `ring_hash` load balancing policy for consistent hashing. This
+policy currently requires using xDS for configuration because users have no
+other way to configure the hash for a request but to use the route configuration
+`hash_policy` field in the `RouteAction` route configuration. This makes the
+`ring_hash` policy unusable without an xDS infrastructure in place.
+
+This proposal extends the configuration of `ring_hash` policy to specify a
+metadata key. The value associated with this metadata key will be used as the
+request hash key if present. This will make it possible to use `ring_hash` by
+configuring it entirely in the [service config][service-config]. If this
+configuration is omitted, we will preserve the current behavior of using the xDS
+hash policy.
+
+### Using an explicit endpoint hash key
+
+Another limitation of the current `ring_hash` load balancing policy is that it
+always hashes the endpoint IP address to place the endpoints on the ring. In
+some scenario, this choice is not ideal: for example, [Kubernetes
+Statefulsets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
+offer a way to configure workloads with sticky identity such that endpoints keep
+storage and network identifier across restarts. However, the IP address may
+change across restarts. After a deployment, if all IP addresses have changed,
+then a completely new ring has to be constructed, even though it may have been
+desirable to keep the ring unchanged based on the Statefulsets identities, so
+that each instance stays at the same location on the ring.
+
+Envoy offers a solution to control endpoint hashes independently of IP
+addresses. This mechanism uses the `"envoy.lb"`
+[LbEndpoint.Metadata][LbEndpoint.Metadata] field `hash_key` value available in
+EDS instead of the endpoint IP address, as described in [the Envoy documentation
+for ring hash][envoy-ringhash].  This proposal adds support for setting the
+endpoint hash key explicitly via EDS by reusing the configuration mechanism
+implemented in Envoy. To retain the advantage of being able to use `ring_hash`
+without xDS, custom gRPC name resolvers will be able to set this endpoint
+attribute through the language specific resolver attribute interface.
+
+### Related Proposals:
+
+This proposal extends the following existing gRFC:
+
+* [gRFC A42: xDS Ring Hash LB Policy][A42]
+
+## Proposal
+
+### Explicitly setting the request hash key
+
+A new string field `request_metadata_key` will be added to the `ring_hash`
```

Comment:
I would like to understand this suggestion. What's the rational for preferring header rather than metadata in this context? The difference between the two is not 100% clear to me. For clients, headers and metadata are basically the same, no?

(I did make the change though)

Reply:
In gRPC, the term "metadata" refers generically to either headers or trailers (sometimes referred to as "initial metadata" and "trailing metadata" respectively).  Saying "headers" makes it clearer which one we mean.  (It couldn't possibly be anything else in this case, since we don't send trailers on the client side, but in general I think it's better for the language to be clearer to avoid any possible confusion.)

In addition, since this gRFC also refers to xDS endpoint metadata, which is something completely different than gRPC metadata, it seems better to use different terms to avoid confusion.